libraries/js/effekt_runtime.js

```diff
@@ -1,385 +1,208 @@
-    handleMonadic: handleMonadic,
-    ref: Cell,
-    state: withState,
-    shift: shift,
-    _if: (c, thn, els) => c ? thn() : els(),
-    withRegion: withRegion,
-
-
-    // Direct style API
-    // ----------------
-    fresh: function(init) {
-      return currentRegion.fresh(init)
-    },
-
-    freshPrompt: function() { return ++_prompt; },
-
-    suspend: function(prompt, body) {
-      _stacksize = 0;
-      throw new Suspension(prompt, body, Nil, Empty)
-    },
-    suspend_bidirectional: function(prompt, caps, body) {
-      throw new Suspension(prompt, body, Cons(thunk => thunk.apply(null, caps), Nil), Empty)
-    },
-
-    // suspension: the raised exception.
-    push: function(suspension, frame) {
-      if (!(suspension instanceof Suspension)) throw suspension;
-      // Assuming `suspension` is a value or variable you want to return
-      throw new Suspension(suspension.prompt, suspension.body,
-        Cons(frame, suspension.frames), suspension.cont);
-    },
-
-    handle: function(prompt, s) {
-      return handleOrRethrow(prompt, s, Nil)
-    },
-
-    freshRegion: function() {
-      return enterRegion(new Arena)
-    },
-
-    leaveRegion: leaveRegion,
-
-    global: global
-  }
-})()
-
-Object.assign($effekt, $runtime);
+// trampolines the given computation (like RUN_TOPLEVEL, but doesn't provide continuations)
+function TRAMPOLINE(comp) {
+  let a = comp;
+  try {
+    while (true) {
+      a = a()
+    }
+  } catch (e) {
+    if (e.computationIsDone) return e.result
+    else throw e
+  }
+}
+
+// keeps the current trampoline going and dispatches the given task
+function RUN(task) {
+  return () => task(TOPLEVEL_KS, TOPLEVEL_K)
+}
+
+// aborts the current continuation
+function ABORT(value) {
+  throw { computationIsDone: true, result: value }
+}
+
+
+// "Public API" used in FFI
+// ------------------------
+
+/**
+ * Captures the current continuation as a WHOLE and makes it available
+ * as argument to the passed body. For example:
+ *
+ *   $effekt.capture(k => ... k(42) ...)
+ *
+ * The body
+ *
+ *   $effekt.capture(k => >>> ... <<<)
+ *
+ * conceptually runs on the _native_ JS call stack. You can call JS functions,
+ * like `setTimeout` etc., but you are not expected or required to return an
+ * Effekt value like `$effekt.unit`. If you want to run an Effekt computation
+ * like in `io::spawn`, you can use `$effekt.run`.
+ *
+ * Advanced usage details:
+ *
+ * The value returned by the function passed to `capture` returns
+ * - a function: the returned function will be passed to the
+ *   Effekt runtime, which performs trampolining.
+ *   In this case, the Effekt runtime will keep running, though the
+ *   continuation has been removed.
+ *
+ * - another value (like `undefined`): the Effekt runtime will terminate.
+ */
+$effekt.capture = CAPTURE
+
+/**
+ * Used to call Effekt function arguments in the JS FFI, like in `io::spawn`.
+ *
+ * Requires an active Effekt runtime (trampoline).
+ */
+$effekt.run = RUN
```

Comment:
I changed the API once more. It now trades-off bare performance (there might be a few more additional runtime checks now) for easy of use.

There now is only _one_ continuation capturing primitive for `async` FFI functions (see docs in `effekt_runtime.js`).